### PR TITLE
[AArch64] Don't tail call memset if it would convert to a bzero.

### PR DESCRIPTION
**Pull request by aemerson**
Well, not quite that simple. We can tc memset since it returns the first
argument but bzero doesn't do that and therefore we can end up miscompiling.

rdar://131419786

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/no-tail-call-bzero-from-memset.ll [hunk 1]
```diff
@@ -0,0 +1,20 @@
+; RUN: llc -o - %s | FileCheck %s
+; RUN: llc -global-isel -o - %s | FileCheck %s
+target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128"
+target triple = "arm64-apple-macosx15.0.0"
+
+define ptr @test()  {
+; CHECK-LABEL: test:
+; CHECK-NOT: b _bzero
```
Comment (efriedma-quic) [inline llvm/test/CodeGen/AArch64/no-tail-call-bzero-from-memset.ll:8]: I'd prefer a positive check for "bl _bzero",

**Review by jroelofs — APPROVED** [review]

File: llvm/lib/CodeGen/Analysis.cpp [hunk 2]
```diff
@@ -685,7 +687,10 @@ bool llvm::returnTypeIsEligibleForTailCall(const Function *F,
          (IID == Intrinsic::memmove &&
           TLI.getLibcallName(RTLIB::MEMMOVE) == StringRef("memmove")) ||
          (IID == Intrinsic::memset &&
-          TLI.getLibcallName(RTLIB::MEMSET) == StringRef("memset"))) &&
+          TLI.getLibcallName(RTLIB::MEMSET) == StringRef("memset") &&
+          (!isa<ConstantInt>(Call->getOperand(1)) ||
+           !cast<ConstantInt>(Call->getOperand(1))->isZero() ||
+           !TLI.getLibcallName(RTLIB::BZERO)))) &&
```
Comment (efriedma-quic) [inline llvm/lib/CodeGen/Analysis.cpp:693]: This is sort of ugly... if this code doesn't precisely match SelectionDAG::getMemset, we miscompile.

Can we rearrange the code so that the caller of isInTailCallPosition() does this computation?  That could be directly adjacent to the code in getMemset that actually makes the decision of whether to call memset or bzero.  We then just pass that into isInTailCallPosition() as a argument.

Comment (aemerson) [reply llvm/lib/CodeGen/Analysis.cpp:693]: Maybe I'm not understanding what you're suggesting. `isInTailCallPosition` is called in multiple places so adding in an extra parameter doesn't make it any cleaner IMO. getMemset() is called later, by that point we've already decided to tail-call. We could downgrade the tail call in getMemset() but then we also lose the tail-call when the parent function's return type was void or it returned undef.

Comment (efriedma-quic) [reply llvm/lib/CodeGen/Analysis.cpp:693]: Currently, isTailCall is a parameter to SelectionDAG::getMemset(), but it doesn't really need to be; SelectionDAGBuilder::visitIntrinsicCall just calls isInTailCallPosition so it can pass the result into getMemset().  Instead, we can just call isInTailCallPosition from getMemset(), once we know whether we're calling bzero or memset, and we can pass that information down.

Comment (aemerson) [reply llvm/lib/CodeGen/Analysis.cpp]: I tried to implement this. It didn't turn out as well as I'd hoped, memset is ok to do and we the logic for memset. However for memmove we have a user in XCore which is computing a tail-call-position using a DAG node without any reference to a `CallInst` so we can't do the same thing there. From what I could tell memcpy usages seemed ok to convert too but I didn't check all of them since there were so many.

File: llvm/lib/CodeGen/Analysis.cpp [hunk 3]
```diff
@@ -739,6 +742,16 @@ bool llvm::returnTypeIsEligibleForTailCall(const Function *F,
   return true;
 }
 
+bool llvm::funcReturnsFirstArgOfCall(const CallInst &CI) {
+    const ReturnInst *Ret = dyn_cast<ReturnInst>(CI.getParent()->getTerminator());
+    Value *RetVal = Ret ? Ret->getReturnValue() : nullptr;
+    bool ReturnsFirstArg = false;
+    if (RetVal && ((RetVal == CI.getArgOperand(0) ||
+                    isPointerBitcastEqualTo(RetVal, CI.getArgOperand(0)))))
```
Comment (jroelofs) [inline llvm/lib/CodeGen/Analysis.cpp:750]: I remember you mentioned the `ret ptr undef` case yesterday. Worth accounting for that here?

File: llvm/lib/CodeGen/Analysis.cpp (hunk 2 above)
Comment (efriedma-quic) [reply llvm/lib/CodeGen/Analysis.cpp]: Due to the way calls are lowered, we have two different tail-call-position analysis passes, yes: one for regular calls, which are lowered as part of building the SelectionDAG, and one for library calls which are created as part of legalization/optimization/etc.

We could teach getMemmove() etc. to call the correct isInTailCallPosition() based on the arguments: if the caller passed in an Instruction, use the IR version, if the caller passed in an SDNode, use the SelectionDAG version.  Or we could try to use the SelectionDAG version exclusively: add an ISD::MEMMOVE node, and lower it after SelectionDAGBuilder.

File: llvm/include/llvm/CodeGen/Analysis.h [hunk 4]
```diff
@@ -144,7 +145,16 @@ bool attributesPermitTailCall(const Function *F, const Instruction *I,
 /// optimization.
 bool returnTypeIsEligibleForTailCall(const Function *F, const Instruction *I,
                                      const ReturnInst *Ret,
-                                     const TargetLoweringBase &TLI);
+                                     const TargetLoweringBase &TLI,
+                                     bool ReturnsFirstArg = false);
+
+/// Check whether B is a bitcast of a pointer type to another pointer type,
+/// which is equal to A.
+bool isPointerBitcastEqualTo(const Value *A, const Value *B);
```
Comment (efriedma-quic) [inline llvm/include/llvm/CodeGen/Analysis.h:153]: Is `isPointerBitcastEqualTo` really necessary?  With opaque pointer types, we should never see a bitcast like this.

File: llvm/lib/CodeGen/Analysis.cpp (hunk 2 above)
Comment (aemerson) [reply llvm/lib/CodeGen/Analysis.cpp]: I think only the IR version tries to do the returns-first-arg trickery so we can probably just go with the first option.

File: llvm/include/llvm/CodeGen/Analysis.h (hunk 4 above)
Comment (aemerson) [reply llvm/include/llvm/CodeGen/Analysis.h:153]: Yeah I guess not.

**Comment by aemerson** [comment]
I converted the other mem routines so we can completely remove the logic in `returnTypeIsEligibleForTailCall()`

**Review by efriedma-quic — APPROVED** [review]
LGTM

File: llvm/include/llvm/CodeGen/Analysis.h [hunk 5]
```diff
@@ -148,10 +148,6 @@ bool returnTypeIsEligibleForTailCall(const Function *F, const Instruction *I,
                                      const TargetLoweringBase &TLI,
                                      bool ReturnsFirstArg = false);
 
-/// Check whether B is a bitcast of a pointer type to another pointer type,
-/// which is equal to A.
-bool isPointerBitcastEqualTo(const Value *A, const Value *B);
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/Analysis.h]: Removing this is an unrelated patch?

Comment (aemerson) [reply llvm/include/llvm/CodeGen/Analysis.h]: Eli commented in another part of the patch that given we have opaque pointers now this isn't needed anymore.
